lib/org/chromium/webidl/Diff.js

```diff
@@ -0,0 +1,247 @@
+        defns[key] = defns[key][0];
+        defns[key].sources = sources;
+      },
+    },
+    {
+      documentation: `Given id (type: { name: String, path: String }),
+        status (type: DiffStatus), and left and right, a DiffChunk is created.
+        If opt_changes is provided, the DiffChunk is placed in opt_changes.
+        Otherwise, it is placed in the array of chunks that is returned.`,
+      name: 'addChunk_',
+      code: function(id, status, left, right, opt_changes) {
+        var chunk = this.DiffChunk.create({
+          definitionName: id.name,
+          status: status,
+          propPath: id.path,
+          leftValue: left,
+          leftSources: id.leftSources,
+          rightValue: right,
+          rightSources: id.rightSources,
+        });
+
+        if (opt_changes) opt_changes.push(chunk);
+        else this.chunks.push(chunk);
+      },
+    },
+    {
+      documentation: `If left and right are of primative types, then a direct
+        comparison will be performed and a DiffChunk is created. If left and
+        right are both arrays, a comparison will be performed between the
+        items. If left and right are objects, it will attempt to recursively
+        traverse properties common to both objects. Traversal ends when a
+        difference is found or there is no more properties to compare.`,
+      name: 'diff_',
+      returns: '',
+      code: function (left, right, id, opt_changes) {
+        if (left === undefined || left === null ||
+            right === undefined || right === null) {
+          if (left !== right)
+            this.addChunk_(id, this.DiffStatus.MISSING_DEFINITION, left,
+                right, opt_changes);
+          return;
+        }
+
+        // Determine types and whether a direct comparison can be done.
+        var leftType = typeof(left);
+        var rightType = typeof(right);
+        var isLeftArray = Array.isArray(left);
+        var isRightArray = Array.isArray(right);
+        foam.assert(leftType === rightType && isLeftArray === isRightArray,
+            'Diff: Encountered a type mismatch!');
+
+        // Type is a basic type, perform simple comparison.
+        if (['boolean', 'number', 'string'].includes(leftType)) {
+          if (left !== right)
+            this.addChunk_(id, this.DiffStatus.VALUES_DIFFER, left,
+                right, opt_changes);
+          return;
+        }
+
+        // Left and right are either Array or Objects at this point.
+        if (isLeftArray) {
+          this.arrayCmp_(left, right, id, opt_changes);
+        } else {
+          foam.assert(left.cls_ && right.cls_,
+              'Diff: Cannot find class property of object.');
+
+          var leftProps = left.cls_.getAxiomsByClass(foam.core.Property);
+          var rightProps = right.cls_.getAxiomsByClass(foam.core.Property);
+          
+          // Determine if left and right have the same properties.
+          if (foam.util.compare(leftProps, rightProps) !== 0) {
```

Comment:
I decided to go with this route because I'm not sure how useful the chunks would be if you were to perform deep comparisons on collections of properties that are different. I would assume you would get a lot of chunks containing `undefined` on one side.